miner/proposal_builder_test.go

```diff
@@ -244,6 +244,121 @@ type step struct {
 	expectErr       string
 }
 
+// The test verifies that the proposal builder creates proposals as soon as possible
+// for initialized and eligible signers, especially if the uninitialized ones take
+// long to init or are blocked entirely.
+func TestBuild_BlockedSignerInitDoesntBlockEligible(t *testing.T) {
+	var signers []*signing.EdSigner
+	rng := rand.New(rand.NewSource(10101))
+	for range 2 {
+		signer, err := signing.NewEdSigner(signing.WithKeyFromRand(rng))
+		require.NoError(t, err)
+		signers = append(signers, signer)
+	}
+
+	var (
+		ctrl      = gomock.NewController(t)
+		conState  = mocks.NewMockconservativeState(ctrl)
+		clock     = mocks.NewMocklayerClock(ctrl)
+		publisher = pmocks.NewMockPublisher(ctrl)
+		trtl      = mocks.NewMockvotesEncoder(ctrl)
+		syncer    = smocks.NewMockSyncStateProvider(ctrl)
+		db        = statesql.InMemoryTest(t)
+		localdb   = localsql.InMemoryTest(t)
+		atxsdata  = atxsdata.New()
+		// singer[1] is blocked
+		atxSearch = mocks.NewMockatxSearch(ctrl)
+	)
+	opts := []Opt{
+		WithLayerPerEpoch(types.GetLayersPerEpoch()),
+		WithLayerSize(2),
+		WithLogger(zaptest.NewLogger(t)),
+		WithSigners(signers...),
+		withAtxSearch(atxSearch),
+	}
+	builder := New(clock, db, localdb, atxsdata, publisher, trtl, syncer, conState, opts...)
+	lid := types.LayerID(15)
+
+	// only signer[0] has ATX
+	atx := gatx(types.ATXID{1}, lid.GetEpoch()-1, signers[0].NodeID(), 1, genAtxWithNonce(777))
+	require.NoError(t, atxs.Add(db, atx, types.AtxBlob{}))
+	atxsdata.AddFromAtx(atx, false)
+	atxSearch.EXPECT().GetIDByEpochAndNodeID(gomock.Any(), lid.GetEpoch()-1, signers[0].NodeID()).DoAndReturn(
+		func(_ context.Context, epoch types.EpochID, nodeID types.NodeID) (types.ATXID, error) {
+			return atxs.GetIDByEpochAndNodeID(db, epoch, nodeID)
+		},
+	)
+	atxSearch.EXPECT().GetIDByEpochAndNodeID(gomock.Any(), lid.GetEpoch()-1, signers[1].NodeID()).DoAndReturn(
+		func(ctx context.Context, epoch types.EpochID, nodeID types.NodeID) (types.ATXID, error) {
+			<-ctx.Done()
+			return types.EmptyATXID, ctx.Err()
+		},
+	)
```

Comment:
I feel like this abstraction is not good and only serves to write a test that I think can't happen.

The context that is passed to `GetIDByEpochAndNodeID` is a) the application context and isn't `Done()` until the application shuts down so a real "stuck" query will be stuck forever.

Also the actual implementation of that interface ignores the context anyway and even if a proper context would be passed (e.g. from `errgroup.WithContext` or a context limited to the end of the current layer, etc.). So the implementation behaves differently from what is tested.